Source/CharacteristicNotificationManager.swift

```diff
@@ -0,0 +1,98 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Polidea
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import CoreBluetooth
+import RxSwift
+
+class CharacteristicNotificationManager {
+
+    private unowned let peripheral: CBPeripheral
+    private let delegateWrapper: CBPeripheralDelegateWrapper
+    private var uuidToActiveObservableMap: [CBUUID: Observable<Characteristic>] = [:]
+    private let lock = NSLock()
+
+    init(peripheral: CBPeripheral, delegateWrapper: CBPeripheralDelegateWrapper) {
+        self.peripheral = peripheral
+        self.delegateWrapper = delegateWrapper
+    }
+
+    func observeValueUpdateAndSetNotification(for characteristic: Characteristic) -> Observable<Characteristic> {
+        return .deferred { [weak self] in
+            guard let strongSelf = self else { throw BluetoothError.destroyed }
+            strongSelf.lock.lock(); defer { strongSelf.lock.unlock() }
+
+            if let activeObservable = strongSelf.uuidToActiveObservableMap[characteristic.uuid] {
+                return activeObservable
+            }
+
+            let notifyValue = strongSelf.setNotifyValue(true, for: characteristic).asObservable()
+            let notificationObserable = strongSelf.createValueUpdateObservable(for: characteristic)
+            let observable = Observable.of(notifyValue, notificationObserable)
+                .merge()
+                .do(onDispose: { [weak self] in
+                    guard let strongSelf = self else { return }
+                    do { strongSelf.lock.lock(); defer { strongSelf.lock.unlock() }
+                        strongSelf.uuidToActiveObservableMap.removeValue(forKey: characteristic.uuid)
```

Comment:
I am just thinking of situation when:
1. Thread 1 stops at line 53
2. Thread 2 locks at line 42 and taking observable on line 44-45
3. Thread 1 removing observable from map at line 55
In this screnario Thread 2 receives observable that is disposed. Let's talk about this when you will be at work